File: python/word_search.py
class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __eq__(self, other):
        return self.x == other.x and self.y == other.y


class WordSearch:
    def __init__(self, puzzle):
        self.puzzle = puzzle

    def search(self, word):
        row = len(self.puzzle)
        col = len(self.puzzle[0])
        word_len = len(word)
        for i in range(row):
            for j in range(col):
                if self.puzzle[i][j] != word[0]:
                    continue
                arr = [
                    (-1, -1),
                    (-1, 0),
                    (-1, 1),
                    (0, 1),
                    (1, 1),
                    (1, 0),
                    (1, -1),
                    (0, -1),
                ]
                for k in range(1, word_len):
                    new_arr = []
                    for dx, dy in arr:
                        nx = i + dx * k
                        ny = j + dy * k
                        if (
                            nx >= 0
                            and ny >= 0
                            and nx < row
                            and ny < col
                            and self.puzzle[nx][ny] == word[k]
                        ):
                            new_arr.append((dx, dy))
                    arr = new_arr
                if arr != []:
                    dx, dy = arr[0]
                    return Point(j, i), Point(
                        j + dy * (word_len - 1), i + dx * (word_len - 1)
                    )

File: elixir/word_search.ex
defmodule WordSearch do
  defmodule Location do
    defstruct [:from, :to]

    @type t :: %Location{
            from: %{row: integer, column: integer},
            to: %{row: integer, column: integer}
          }

    def new({x1, y1}, {x2, y2}) do
      %Location{from: %{row: x1, column: y1}, to: %{row: x2, column: y2}}
    end
  end

  @doc """
  Find the start and end positions of words in a grid of letters.
  Row and column positions are 1 indexed.
  """
  @spec search(grid :: String.t(), words :: [String.t()]) :: %{String.t() => nil | Location.t()}
  def search(grid, words) do
    mp =
      grid
      |> String.split("\n")
      |> Enum.with_index(1)
      |> Enum.reduce(%{}, fn {r, i}, acc ->
        r
        |> String.graphemes()
        |> Enum.with_index(1)
        |> Enum.reduce(acc, fn {c, j}, acc ->
          Map.put(acc, {i, j}, c)
        end)
      end)

    words
    |> Map.new(fn word ->
      ws = String.graphemes(word)

      mp
      |> Enum.find_value(fn {p, _c} ->
        [{-1, -1}, {-1, 0}, {-1, 1}, {0, 1}, {1, 1}, {1, 0}, {1, -1}, {0, -1}]
        |> Enum.find_value(fn d -> check(mp, p, d, ws) end)
        |> case do
          nil -> nil
          q -> Location.new(p, q)
        end
      end)
      |> then(&{word, &1})
    end)
  end

  defp check(_, {x, y}, {dx, dy}, []), do: {x - dx, y - dy}

  defp check(mp, curr = {x, y}, d = {dx, dy}, [w | ws]) do
    if mp[curr] != w do
      nil
    else
      check(mp, {x + dx, y + dy}, d, ws)
    end
  end
end

